### ISSUE TITLE
Chinese search in version 2.2.6 has problems and cannot be found

### ISSUE THREAD

**Issue by Rhyheart** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#885]
Chinese search in version 2.2.6 has problems and cannot be found，It seems that this problem only exists in 2.2.6. When I demote to 2.2.0, there will be no problem.

**Comment by Rhyheart** [comment]
It seems that this problem only exists in 2.2.6. When I demote to 2.2.0, there will be no problem.

**Comment by lauxjpn** [comment]
Please provide us with some source code, that you try to run, the result that you are expecting and the actual result that you are getting.

**Comment by Rhyheart** [comment]
![image](https://user-images.githubusercontent.com/44357526/67147526-46664800-f2c8-11e9-8515-d9c91c75d2ed.png)
The problem is described in the screenshot

**Comment by lauxjpn** [comment]
Thank you for clearing this up. I will try to reproduce this issue.

**Comment by lauxjpn** [comment]
I cannot reproduce this issue on 3.0.0.

The following code works as expected:

```c#
using System.Linq;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.TestUtilities;
using Pomelo.EntityFrameworkCore.MySql.FunctionalTests.TestUtilities;
using Xunit;

namespace Pomelo.EntityFrameworkCore.MySql.FunctionalTests
{
    public class ReportedIssuesMySqlTest
        : IClassFixture<ReportedIssuesMySqlTest.ReportedIssuesMySqlFixture>
    {
        [ConditionalFact]
        public virtual void TestIssue885()
        {
            using var context = CreateContext();

            var queryEnabledNegations = context.Set<OpenDream>()
                .Where(e => e.IsEnabled && e.Keywords.Contains("無"))
                .ToArray();

            Assert.Equal(2, queryEnabledNegations.Length);
        }

        public class OpenDream
        {
            public int OpenDreamId { get; set; }
            public bool IsEnabled { get; set; }
            public string Keywords { get; set; }
        }

        public class ReportedIssuesMySqlFixture : SharedStoreFixtureBase<PoolableDbContext>
        {
            protected override void OnModelCreating(
                ModelBuilder modelBuilder,
                DbContext context)
            {
                modelBuilder.Entity<OpenDream>(entity =>
                {
                    entity.Property(e => e.Keywords)
                        .HasMaxLength(255);
                });
            }

            protected override void Seed(PoolableDbContext context)
            {
                context.Set<OpenDream>().AddRange(
                    new OpenDream
                    {
                        IsEnabled = true,
                        Keywords = "職名技海権彫額般将物月直新埋宅喪上著仏掲。"
                    },
                    new OpenDream
                    {
                        IsEnabled = true,
                        Keywords = "治統覧劇提一正村柳野速将無管遺覧資読記雅。"
                    },
                    new OpenDream
                    {
                        IsEnabled = false,
                        Keywords = "通第生場無一誰偶助掲商幅掲賛。"
                    },
                    new OpenDream
                    {
                        IsEnabled = true,
                        Keywords = "将無裁魚写大文責的時内刑答実郷断攻活間楽。"
                    }
                );

                context.SaveChanges();
            }

            protected override ITestStoreFactory TestStoreFactory
                => MySqlTestStoreFactory.Instance;

            protected override string StoreName { get; } = "Issue885";
        }

        public ReportedIssuesMySqlTest(ReportedIssuesMySqlFixture fixture)
            => Fixture = fixture;

        protected ReportedIssuesMySqlFixture Fixture { get; }
        protected DbContext CreateContext() => Fixture.CreateContext();
    }
}
```

It generates the following SQL:

```sql
CREATE DATABASE `Issue885`;

CREATE TABLE `OpenDream` (
    `OpenDreamId` int NOT NULL AUTO_INCREMENT,
    `IsEnabled` tinyint(1) NOT NULL,
    `Keywords` varchar(255) CHARACTER SET utf8mb4 NULL,
    CONSTRAINT `PK_OpenDream` PRIMARY KEY (`OpenDreamId`)
);

@p0='True'
@p1='職名技海権彫額般将物月直新埋宅喪上著仏掲。' (Size = 255)
@p2='True'
@p3='治統覧劇提一正村柳野速将無管遺覧資読記雅。' (Size = 255)
@p4='False'
@p5='通第生場無一誰偶助掲商幅掲賛。' (Size = 255)
@p6='True'
@p7='将無裁魚写大文責的時内刑答実郷断攻活間楽。' (Size = 255)

INSERT INTO `OpenDream` (`IsEnabled`, `Keywords`)
VALUES (@p0, @p1);
SELECT `OpenDreamId`
FROM `OpenDream`
WHERE ROW_COUNT() = 1 AND `OpenDreamId` = LAST_INSERT_ID();

INSERT INTO `OpenDream` (`IsEnabled`, `Keywords`)
VALUES (@p2, @p3);
SELECT `OpenDreamId`
FROM `OpenDream`
WHERE ROW_COUNT() = 1 AND `OpenDreamId` = LAST_INSERT_ID();

INSERT INTO `OpenDream` (`IsEnabled`, `Keywords`)
VALUES (@p4, @p5);
SELECT `OpenDreamId`
FROM `OpenDream`
WHERE ROW_COUNT() = 1 AND `OpenDreamId` = LAST_INSERT_ID();

INSERT INTO `OpenDream` (`IsEnabled`, `Keywords`)
VALUES (@p6, @p7);
SELECT `OpenDreamId`
FROM `OpenDream`
WHERE ROW_COUNT() = 1 AND `OpenDreamId` = LAST_INSERT_ID();

SELECT `o`.`OpenDreamId`, `o`.`IsEnabled`, `o`.`Keywords`
FROM `OpenDream` AS `o`
WHERE `o`.`IsEnabled` AND (LOCATE(BINARY '無', `o`.`Keywords`) > 0)
```

I used the following `DbContextOptionsBuilder` options:

```c#
builder.UseMySql(ConnectionString,
    b => b.CharSetBehavior(CharSetBehavior.AppendToAllColumns)
          .AnsiCharSet(CharSet.Utf8mb4)
          .UnicodeCharSet(CharSet.Utf8mb4)
);
```

Please provide us with:
- The `CREATE TABLE` SQL statement of your `OpenDream` table
- The `DbContextOptionsBuilder` calls (`UseMySql()` and options)
- The `SELECT` SQL query, that is generated when calling `x.KeyWord.Contains()`

**Comment by yukozh** [comment]
Please open an issue by following the issue template, without repro steps, expected behaviors, actual behaviors will waste lots of time on communication.

**Comment by lauxjpn** [comment]
@Rhyheart As for the actual issue, I think it it is very likely, that the default charset of your database and the charset of the column `KeyWord` (the column you are checking) do mismatch. If that is the case, the generated SQL might not return the rows you would expect, because we use the `BINARY` keyword since 2.2.6 to force the check to be case sensitive.

When I implemented this fix, I did not realize however, that this might be a breaking change for someone and because we follow SemVer, we should introduce breaking changes only in major versions.

We have a couple of options here:

1. We could alter the new behavior in 2.2.6, so that it is not using the `BINARY` statement anymore, but instead forces the use of a case sensitive collation that is based on the `AnsiCharSet()` and `UnicodeCharSet()` extensions.

2. We could just remove the original fix, skip the then failing test and keep the behavior in 2.2.6 as it was before. This would only be introduced in 3.0.0 then and therefore could be a breaking change if necessary (though we might want to implement it there as in the previous paragraph as well).

3. There is also the possibility to introduce a `HasCollation()` property builder extension method. This could also be used in conjunction with the first option.

**Comment by Rhyheart** [comment]
![image](https://user-images.githubusercontent.com/44357526/67390173-b0cffe80-f5ce-11e9-9baa-2fa73cead085.png)
Thank you very much. I'll add this code, and then it's all right.

**Comment by lauxjpn** [comment]
This issue is fixed for `2.2.6` with #891 (on the `2.2-maint` branch). The workaround will not be needed then anymore (but of course does no harm either).

**Comment by Rhyheart** [comment]
> This issue is fixed for `2.2.6` with #891 (on the `2.2-maint` branch). The workaround will not be needed then anymore (but of course does no harm either).

Thank you very much. I'll upgrade it when you release it.

**Comment by AlphaGit** [comment]
Hey @lauxjpn -- thanks for taking care of this problem, which I found too.

From your comment, it wasn't entirely clear to me: is the fix going to be present in 2.2.6 to preserve SemVer behavior?

Or is the fix going to be post-2.2.6? (maybe 2.2.7? 3.0?)

**Comment by lauxjpn** [comment]
> This issue is fixed for 2.2.6 with #891 (on the 2.2-maint branch). The workaround will not be needed then anymore (but of course does no harm either).

The issue is already fixed on the `2.2-maint` branch and will be part of future release there (the actual name of the next version will either be `2.2.7`, or (maybe more in line with our policy to name our versions according to the EF Core version) something like `2.2.6-sr1` (`sr` meaning `service release`); what do you think @caleblloyd?). It should currently only be part of the `2.2-maint` nightly builds.

For the `3.0.0` release, this issue is tracked by #890 and will be part of `3.0.0-rc2`, which will be released on Monday.
